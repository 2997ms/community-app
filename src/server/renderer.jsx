/**
 * This module implements ExpressJS middleware for server-side rendering of
 * the App.
 */

import config from 'config';
import React from 'react';
import ReactDOM from 'react-dom/server';
import { StaticRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import serializeJs from 'serialize-javascript';
import { isDev } from 'utils/isomorphy';

import App from '../shared';

/* This is always initial state of the store. Later we'll have to provide a way
 * to put the store into correct state depending on the demanded route. */
import storeFactory from '../shared/store-factory';

export default (req, res) => {
  storeFactory(req).then((store) => {
    const context = {};
    const appHtml = ReactDOM.renderToString((
      <Provider store={store}>
        <StaticRouter
          context={context}
          location={req.url}
        >
          <App />
        </StaticRouter>
      </Provider>
    ));
    if (context.status) res.status(context.status);
    res.send((
      `<!DOCTYPE html>
      <html>
        <head>
          <title>Topcoder</title>
          ${
            /* In dev environment styles are embed into JS bundle,
              and we should not include this link into the page, otherwise
              the page will initially load production /style.css and it may
              result in some wierd behavior while loading. */
            isDev() ? '' : '<link rel="stylesheet" href="/style.css" />'
          }
        </head>
        <body>
          <div id="react-view">${appHtml}</div>
          <script type="application/javascript">
            window.CONFIG = ${serializeJs(config, { isJSON: true })}
            window.ISTATE = ${serializeJs(store.getState(), { isJSON: true })}
          </script>
          <script type="application/javascript" src="/bundle.js"></script>
        </body>
      </html>`
    ));
  });
};
