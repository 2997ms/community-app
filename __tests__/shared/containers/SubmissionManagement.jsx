import _ from 'lodash';
import React from 'react';
import renderer from 'react-test-renderer';
import TU from 'react-dom/test-utils';

const mockChallengeActions = {
  fetchChallengeInit: jest.fn(),
  fetchChallengeDone: jest.fn(),
  fetchSubmissionsInit: jest.fn(),
  fetchSubmissionsDone: jest.fn(),
};
jest.setMock(require.resolve('actions/challenge'), mockChallengeActions);

const mockSmpActions = {
  smp: {
    cancelDelete: jest.fn(),
    confirmDelete: jest.fn(),
    deleteSubmissionDone: jest.fn(),
    downloadSubmission: jest.fn(),
    showDetails: jest.fn(),
  },
};
jest.setMock(require.resolve('actions/smp'), mockSmpActions);

const mockState = {
  auth: {},
  challenge: {
    details: {
      track: 'Track',
    },
    loadingDetails: true,
    mySubmissionsManagement: {
      showDetails: new Set(),
      showModal: true,
    },
  },
};

const SubmissionManagement = require('containers/SubmissionManagement').default;

beforeEach(() => jest.clearAllMocks());

test('Matches shapshot', () => {
  const render = renderer.create((
    <SubmissionManagement
      match={{
        params: {
          challengeId: '12345',
        },
      }}
      store={{
        dispatch: () => _.noop,
        getState: () => mockState,
        subscribe: _.noop,
      }}
    />
  ));
  expect(render.toJSON()).toMatchSnapshot();
});

const obj = TU.renderIntoDocument((
  <SubmissionManagement
    match={{
      params: {
        challengeId: '12345',
      },
    }}
    store={{
      dispatch: () => _.noop,
      getState: () => mockState,
      subscribe: _.noop,
    }}
  />
));
const props = obj.selector.props;

test('onShowDetails dispatches', () => {
  props.onShowDetails('12345');
  expect(mockSmpActions.smp.showDetails).toHaveBeenCalledWith('12345');
});

test('onSubmissionDelete dispatches', () => {
  props.onSubmissionDelete('12345');
  expect(mockSmpActions.smp.confirmDelete).toHaveBeenCalledWith('12345');
});

test('onCancelSubmissionDelete dispatches', () => {
  props.onCancelSubmissionDelete();
  expect(mockSmpActions.smp.cancelDelete).toHaveBeenCalledWith();
});

test('onSubmissionDeleteConfirmed dispatches', () => {
  props.onSubmissionDeleteConfirmed('12345', '54321');
  expect(mockSmpActions.smp.deleteSubmissionDone)
    .toHaveBeenCalledWith('12345', '54321');
});

test('onDownloadSubmission dispatches', () => {
  props.onDownloadSubmission('12345', '54321');
  expect(mockSmpActions.smp.downloadSubmission)
    .toHaveBeenCalledWith('12345', '54321');
});

test('loadChallengeDetails dispatches', () => {
  props.loadChallengeDetails('12345', '54321');
  expect(mockChallengeActions.fetchChallengeInit).toHaveBeenCalled();
  expect(mockChallengeActions.fetchChallengeDone)
    .toHaveBeenCalledWith('12345', '54321');
});

test('loadMySubmissions dispatches', () => {
  props.loadMySubmissions('12345', '54321');
  expect(mockChallengeActions.fetchSubmissionsInit).toHaveBeenCalled();
  expect(mockChallengeActions.fetchSubmissionsDone)
    .toHaveBeenCalledWith('12345', '54321');
});
