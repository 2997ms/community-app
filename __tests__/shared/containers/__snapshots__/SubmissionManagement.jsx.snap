// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches shapshot 1`] = `
<SubmissionManagementPageContainer
  authTokens={
    Object {
      "tokenV3": "Token V3",
    }
  }
  challenge={
    Object {
      "track": "Track",
    }
  }
  challengeId={12345}
  challengesUrl="/challenges"
  deleting={true}
  isLoadingChallenge={false}
  isLoadingSubmissions={false}
  loadChallengeDetails={[Function]}
  loadMySubmissions={[Function]}
  match={
    Object {
      "params": Object {
        "challengeId": 12345,
      },
    }
  }
  mySubmissions={
    Array [
      Object {
        "submissionId": 12345,
      },
    ]
  }
  onCancelSubmissionDelete={[Function]}
  onDownloadSubmission={[Function]}
  onShowDetails={[Function]}
  onSubmissionDelete={[Function]}
  onSubmissionDeleteConfirmed={[Function]}
  showDetails={Set {}}
  showModal={true}
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "notify": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "dispatch": [Function],
        "getState": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": null,
    }
  }
  toBeDeletedId={12345}
/>
`;
